pjrt/convert/trace_container.zig

```diff
@@ -0,0 +1,211 @@
+
+pub const xla_async_op_line_name = "Async XLA Ops";
+
+pub const host_threads_plane_name = "/host:CPU";
+pub const gpu_plane_prefix = "/device:GPU:";
+pub const tpu_plane_prefix = "/device:TPU:";
+pub const custom_plane_prefix = "/device:CUSTOM:";
+
+pub fn GetOrPutResult(comptime Key: type, comptime Value: type) type {
+    return struct {
+        key_ptr: *Key,
+        value_ptr: *Value,
+    };
+}
+
+pub const TraceContainer = struct {
+    metadata: Trace = .{},
+    events: std.ArrayListUnmanaged(*TraceEvent) = .{},
+
+    pub const Trace = struct {
+        devices: std.AutoArrayHashMapUnmanaged(u32, Device) = .{},
+        trace_events: std.ArrayListUnmanaged(TraceEvent) = .{},
+    };
+
+    pub const Device = struct {
+        name: []const u8 = &[_]u8{},
+        device_id: u32 = 0,
+        resources: std.AutoArrayHashMapUnmanaged(u32, trace_events_proto.Resource) = .{},
+    };
+
+    pub const TraceEvent = struct {
+        device_id: u32 = 0,
+        resource_id: u32 = 0,
+        name: []const u8 = &[_]u8{},
+        timestamp_ps: u128 = 0,
+        duration_ps: u64 = 0,
+        args: std.StringArrayHashMapUnmanaged([]const u8) = .{},
+    };
+
+    fn buildDeviceAndResources(allocator: std.mem.Allocator, device_id: u32, plane: *const xplane_visitor.XPlaneVisitor, device: *Device) !void {
+        device.name = plane.name();
+        device.device_id = device_id;
+        const sort_by_ordinal = (device_id == host_threads_device_id);
+        var ordinal: u32 = 0;
+        for (plane.plane.lines.items) |*l| {
+            var line = xplane_visitor.XLineVisitor.init(plane, l);
+            const resource_id: u32 = @intCast(line.displayId());
+            var resource: trace_events_proto.Resource = .{
+                .resource_id = resource_id,
+                .name = .{ .Const = line.displayName() },
+            };
+
+            if (sort_by_ordinal) {
+                ordinal += 1;
+                resource.sort_index = ordinal;
+            }
+            try device.resources.put(allocator, resource_id, resource);
+        }
+    }
+
+    fn xplaneToTraceEvents(self: *TraceContainer, allocator: std.mem.Allocator, device_id: u32, xplane: *const xplane_visitor.XPlaneVisitor) !void {
+        // Convert devices and resources.
+        const device_entry = try self.metadata.devices.getOrPut(allocator, device_id);
+        if (!device_entry.found_existing) device_entry.value_ptr.* = .{};
+
+        try buildDeviceAndResources(allocator, device_id, xplane, device_entry.value_ptr);
+
+        // Convert events.
+        for (xplane.plane.lines.items) |*l| {
+            const xline = xplane_visitor.XLineVisitor.init(xplane, l);
+            const resource_id: u32 = @intCast(xline.displayId());
+            if (std.mem.eql(u8, xline.displayName(), xla_async_op_line_name)) continue;
+            for (xline.line.events.items) |*e| {
+                const xevent = xplane_visitor.XEventVisitor.init(xline.plane, xline.line, e);
```

Comment:
same here with xline. This code is obfuscated, since "xplane" and "l" have become "xline.plane" and "xline.line". With less intermediary objects it would be easier to follow.